### DISCUSSION TITLE
Local applications in window problems

### DISCUSSION THREAD

**Discussion by pcisar** [rio-labs/rio#71] [Q&A] [answered]
According to [Rio setup](https://rio.dev/docs/tutorial-tic-tac-toe/1-rio-setup) documentation, it's possible to run local applications in separate window instead in regular browser. However, there is no explanation how to run such application! One would expect that when rio-ui[window] is installed and project is "App", then `rio run` will run it in window, but that's not the case. The 'App' class has `run_in_window()` method that is called from generated `__main__.py` file for the project. Running `python -m PROJECTPKG` opens the blank window, but that's it. It's hard to tell where the problem is, whether in installed stack or Rio. It seems that Rio depends on rio.toml that is not accessible when project is started in other way than `rio run`.

Another big question is how such local applications could be deployed.  Rio project is not a regular package that could be easily built as wheel. I have no problems to create `pyproject.toml` for it (I use `hatch` for packaging), adjust project layout to my standard `src` style, rename `__main__.py` to `main.py` and wrap the `PROJECT.app.run_in_window()` into `run` function, then make it as script entry point in `pyproject.toml`. The deployment via wheel works nicely, except the empty window problem. However, it would be nice if Rio would spare us from going trough these hoops with better project template for Applications.

**Comment by mad-moo** [comment]
Hey!

We can't reproduce this. After creating an app and starting it with `rio run` the window fires right up, just as it should. Can you go into more details about what steps you've taken? Also, what platform are you on? Window support is still experimental and works best on Windows for now.

What changes would you like to see exactly as far as app templates go? Wouldn't those changes be specific to one particular packaging tool?

**Comment by pcisar** [comment]
OpenSuSE Tumbleweed

-> mkvirtualenv -p3.11 rioapp
-> pip install rio-ui[window]

freeze file:
----
aiofiles==23.2.1
annotated-types==0.7.0
anyio==4.4.0
blessed==1.20.0
bottle==0.12.25
certifi==2024.6.2
cffi==1.16.0
click==8.1.7
colorama==0.4.6
contourpy==1.2.1
copykitten==1.2.0
crawlerdetect==0.1.7
cryptography==42.0.8
cycler==0.12.1
dnspython==2.6.1
email_validator==2.2.0
fastapi==0.111.0
fastapi-cli==0.0.4
fonttools==4.53.0
fuzzywuzzy==0.18.0
gitignore_parser==0.1.11
h11==0.14.0
httpcore==1.0.5
httptools==0.6.1
httpx==0.27.0
idna==3.7
importlib_metadata==8.0.0
imy==0.3.6
introspection==1.9.1
isort==5.13.2
jaraco.classes==3.4.0
jeepney==0.8.0
Jinja2==3.1.4
json5==0.9.25
keyring==24.3.1
kiwisolver==1.4.5
langcodes==3.4.0
language_data==1.2.0
Levenshtein==0.25.1
marisa-trie==1.2.0
markdown-it-py==3.0.0
MarkupSafe==2.1.5
matplotlib==3.9.0
mdurl==0.1.2
more-itertools==10.3.0
motor==3.5.0
multidict==6.0.5
numpy==2.0.0
ordered-set==4.1.0
orjson==3.10.5
packaging==24.1
pillow==10.3.0
pkg-about==1.1.6
platformdirs==3.11.0
proxy-tools==0.1.0
pycparser==2.22
pydantic==2.7.4
pydantic_core==2.18.4
Pygments==2.18.0
pymongo==4.8.0
pyparsing==3.1.2
PyQt5==5.15.10
PyQt5-Qt5==5.15.2
PyQt5-sip==12.13.0
PyQtWebEngine==5.15.6
PyQtWebEngine-Qt5==5.15.2
python-dateutil==2.9.0.post0
python-dotenv==1.0.1
python-Levenshtein==0.25.1
python-multipart==0.0.9
pytz==2024.1
pywebview==4.4.1
PyYAML==6.0.1
QtPy==2.4.1
rapidfuzz==3.9.3
readchar==4.1.0
renumerate==1.1.12
revel==0.9.1
rich==13.7.1
rio-ui==0.9.1
SecretStorage==3.3.3
sentinel==1.0.0
shellingham==1.5.4
six==1.16.0
sniffio==1.3.1
starlette==0.37.2
timer-dict==1.0.0
tomlkit==0.12.5
typer==0.12.3
typing_extensions==4.12.2
ujson==5.10.0
unicall==0.1.5
uniserde==0.3.14
uvicorn==0.29.0
uvloop==0.19.0
watchfiles==0.22.0
wcwidth==0.2.13
websockets==12.0
yarl==1.9.4
zipp==3.19.2

----
->rio new
name: riotest
template: Tic-Tac-Toe
Type: App

----

-> cat rio.toml
> \# This is the configuration file for Rio, an easy to use app & web framework for 
> \# Python.
>
> [app]
> \# This is either "website" or "app"
> app-type = "app"
> \# The name of your Python module
> main-module = "riotest"
> \# All files which are part of your project. Changes to these will trigger a
> \# reload and they will be packed up when deploying.
> project-files = ["*.py", "/assets/", "/rio.toml"]

----

-> rio run

  _____  _
 |  __ \(_)
 | |__) |_  ___
 |  _  /| |/ _ \
 | | \ \| | (_) |
 |_|  \_\_|\___/


Starting...
Warning: Debug mode is enabled.
Warning: Debug mode includes helpful tools for development, but is slower and disables some safety checks. Never use it in production!
Warning: Run with `--release` to disable debug mode.

Running in local mode. Only this device can access the app.

The app is running at http://127.0.0.1:8000

🗨  Join other Developers on Discord — https://discord.gg/7ejXaPwhyH
✨ Star Rio on GitHub — https://github.com/rio-labs/rio
📣 Spread the word

----

The same will happen with `--release`

----

-> cd riotest
-> python -m riotest
Will open the window, but it remains empty. When window is closed, the application ends normally.

**Comment by pcisar** [comment]
Interesting, when using Rio from git repository, the `rio run` opens the window, but it's still blank. However, now the terminal reports an error:
`FileNotFoundError: [Errno 2] Directory or file not exists: '/home/job/python/projects/rio/rio/rio/frontend files/index.html'`

The CWD is `/home/job/python/projects/rio/riotest`

Traceback:
----

> Exception in ASGI application
> Traceback (most recent call last):
>   File "/home/job/python/envs/riotest/lib/python3.11/site-packages/uvicorn/protocols/http/httptools_impl.py", line 411, in run_asgi
>     result = await app(  # type: ignore[func-returns-value]
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/job/python/envs/riotest/lib/python3.11/site-packages/uvicorn/middleware/proxy_headers.py", line 69, in __call__
>     return await self.app(scope, receive, send)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/job/python/projects/rio/rio/rio/app_server/fastapi_server.py", line 250, in __call__
>     return await super().__call__(scope, receive, send)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/job/python/envs/riotest/lib/python3.11/site-packages/fastapi/applications.py", line 1054, in __call__
>     await super().__call__(scope, receive, send)
>   File "/home/job/python/envs/riotest/lib/python3.11/site-packages/starlette/applications.py", line 123, in __call__
>     await self.middleware_stack(scope, receive, send)
>   File "/home/job/python/envs/riotest/lib/python3.11/site-packages/starlette/middleware/errors.py", line 186, in __call__
>     raise exc
>   File "/home/job/python/envs/riotest/lib/python3.11/site-packages/starlette/middleware/errors.py", line 164, in __call__
>     await self.app(scope, receive, _send)
>   File "/home/job/python/envs/riotest/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 65, in __call__
>     await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
>   File "/home/job/python/envs/riotest/lib/python3.11/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
>     raise exc
>   File "/home/job/python/envs/riotest/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
>     await app(scope, receive, sender)
>   File "/home/job/python/envs/riotest/lib/python3.11/site-packages/starlette/routing.py", line 756, in __call__
>     await self.middleware_stack(scope, receive, send)
>   File "/home/job/python/envs/riotest/lib/python3.11/site-packages/starlette/routing.py", line 776, in app
>     await route.handle(scope, receive, send)
>   File "/home/job/python/envs/riotest/lib/python3.11/site-packages/starlette/routing.py", line 297, in handle
>     await self.app(scope, receive, send)
>   File "/home/job/python/envs/riotest/lib/python3.11/site-packages/starlette/routing.py", line 77, in app
>     await wrap_app_handling_exceptions(app, request)(scope, receive, send)
>   File "/home/job/python/envs/riotest/lib/python3.11/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
>     raise exc
>   File "/home/job/python/envs/riotest/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
>     await app(scope, receive, sender)
>   File "/home/job/python/envs/riotest/lib/python3.11/site-packages/starlette/routing.py", line 72, in app
>     response = await func(request)
>                ^^^^^^^^^^^^^^^^^^^
>   File "/home/job/python/envs/riotest/lib/python3.11/site-packages/fastapi/routing.py", line 278, in app
>     raw_response = await run_endpoint_function(
>                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/job/python/envs/riotest/lib/python3.11/site-packages/fastapi/routing.py", line 191, in run_endpoint_function
>     return await dependant.call(**values)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/job/python/projects/rio/rio/rio/app_server/fastapi_server.py", line 388, in _serve_index
>     html_ = read_frontend_template("index.html")
>             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/job/python/projects/rio/rio/rio/app_server/fastapi_server.py", line 102, in read_frontend_template
>     return (utils.FRONTEND_FILES_DIR / template_name).read_text(
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/lib64/python3.11/pathlib.py", line 1058, in read_text
>     with self.open(mode='r', encoding=encoding, errors=errors) as f:
>          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/lib64/python3.11/pathlib.py", line 1044, in open
>     return io.open(self, mode, buffering, encoding, errors, newline)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Comment by pcisar** [comment]
The Rio repository is cloned to `/home/job/python/projects/rio/rio`, installed in editable mode after all requirements. However, the same error happens when Rio is built and installed directly via `pip install .`, just the missing file path is different: `/home/job/python/envs/riotest/lib/python3.11/site-packages/rio/frontend files/index.html`

**Reply by mad-moo** [reply]
You need to build Rio. Have a look at the contribution guidelines if you want to go down that path: https://github.com/rio-labs/rio/blob/dev/CONTRIBUTING.md

Can I ask why you aren't using the prebuilt version from PyPI?

**Reply by pcisar** [reply]
Because the prebuilt version from PyPI (rio-ui==0.9.1) had this problem I reported (app not started in windows via `rio run`). Then you wrote that you couldn't reproduce it, hence I figured that some unreleased commit fixed that. Well, it did, sort of, hence the follow up.

**Comment by Aran-Fey** [comment, marked as answer]
Regarding the empty white window: I've had a similar experience with `run_in_window` on linux. In my case it was because some system dependencies were missing. Please make sure to follow the [pywebview installation instructions](https://pywebview.flowrl.com/guide/installation.html#linux).

There are two different webview backends available on linux, qt and gtk. So even if you've installed all the dependencies, it might be using the wrong backend. Try explicitly telling it which backend to use by setting `export PYWEBVIEW_GUI=qt` or `export PYWEBVIEW_GUI=gtk`.

**Reply by pcisar** [reply]
Thanks, I have managed to run it with GTK. Still did not figured out what makes the Qt version fail. I should have all dependencies installed, but as there is no log or other output, it's hard to tell what's going on here. 

But this issue clearly demonstrates the problems with Rio apps deployment. I like Rio a lot, it clearly has a great potential. Hence I'm willing to close my eyes to incredibly huge dependencies even for "Hello World" app, if  the app deployment to end user would be straightforward, which is not yet. It would be really great if there will be at least some How To style documentation on this topic. Some direct support from Rio for local-app distribution packages would be also welcome. For personal use deployment, one could install Rio, all dependencies  required to run local-apps, then one or more custom Rio apps, create scripts to run `rio run` from  proper locations to start each app. But it's not suitable for installation in business environment.

**Reply by Aran-Fey** [reply]
I can see where you're coming from, and I agree that the documentation around local apps should be improved. But that said, rio's primary focus are web apps. Local apps are unfortunately quite tricky (you've already seen how bad the webview support is on linux), and we currently don't have the time/resources to dive into that topic. It's something we will tackle once rio has matured a bit as a framework.

Have you considered deploying your app on a web server? I think that would be easier than setting it up on every user's PC.

**Reply by pcisar** [reply]
Yes, I saw the code, and local apps are tricky. But support for both web and local apps is one from the advantages of Rio. I understand that Rio is relatively young project, so better local app support will materialize over time. I can start my development with what is available now. I hoped that I would be able to help a bit with some small things once I'll get more familiar with the framework, but your dev. system is complex and somewhat alien to me (I'm not a fan of `rye`, as all my projects are using `hatch`). Anyway, thanks for your time, it really helped me.

**Comment by mad-moo** [comment]
That could well be, but in this case I'd say it's for the same reason as above. Without `index.html`, there is nothing for the browser to show
